### ISSUE TITLE
Segfault when plotting large rectilinear grid

### ISSUE THREAD

**Issue by aowen87** [aowen87/BAR#2699]
We have a user that is trying to visualize a 1536x1536x1297 rectilinear grid, specified using an Xdmf file and two HDF5 files (one for grid coordinates and one for the volume data itself).

Loading in Visit 2.12.2 (official binary) works fine, but two cases cause a segfault, in different places as far as I can tell from the logs:

1. Add Mesh plot -> thetacut; Draw
2. Add Volume plot -> Azimuthal Velocity; disable lighting in the volume properties; Draw

I've added set of log files for both cases. Also a dummy xdmf file using the original grid coordinate HDF5 file and a script to generate dummy volume data (I can't share the data itself).

The viz node I'm running this on has 64G memory, and I don't see the usage getting close to the limit.

-----------------------REDMINE MIGRATION-----------------------
This ticket was migrated from Redmine. As such, not all 
information was able to be captured in the transition. Below is
a complete record of the original redmine ticket.

Ticket number: 2774
Status: Resolved
Project: VisIt
Tracker: Bug
Priority: Normal
Subject: Segfault when plotting large rectilinear grid
Assigned to: Mark Miller
Category: 
Target version: 2.13.1
Author: Paul Melis
Start: 02/24/2017
Due date: 
% Done: 0
Estimated time: 
Created: 02/24/2017 09:22 am
Updated: 03/02/2018 04:25 pm
Likelihood: 3 - Occasional
Severity: 4 - Crash / Wrong Results
Found in version: 2.12.1
Impact: 
Expected Use: 
OS: Linux
Support Group: Any
Description:
 We have a user that is trying to visualize a 1536x1536x1297 rectilinear grid, specified using an Xdmf file and two HDF5 files (one for grid coordinates and one for the volume data itself).

Loading in Visit 2.12.2 (official binary) works fine, but two cases cause a segfault, in different places as far as I can tell from the logs:

1. Add Mesh plot -> thetacut; Draw
2. Add Volume plot -> Azimuthal Velocity; disable lighting in the volume properties; Draw

I've added set of log files for both cases. Also a dummy xdmf file using the original grid coordinate HDF5 file and a script to generate dummy volume data (I can't share the data itself).

The viz node I'm running this on has 64G memory, and I don't see the usage getting close to the limit.

Comments:
The original volume data actually used 64-bit floats, but I've reduced it to 32-bit floats for this bug report.
The change of Likelihood to Occasional suggests there are situations in which this does not occur? Is there any workaround I can give to the user, as he is currently stuck in visualizing this data.
I looked a bit into the stack trace after the segfault (for the case of adding the Volume plot) and it seems to be a case of integer overflow. This is with 2.12.2 sources, btw.The index array is set in avtMassVoxelExtractor::ExtractImageSpaceGrid() around line 2640:
                    int index[8];
                    index[0] = (zind)*nX*nY + (yind)*nX + (xind);
                    index[1] = (zind)*nX*nY + (yind)*nX + (xind+1);
                    index[2] = (zind)*nX*nY + (yind+1)*nX + (xind);
                    index[3] = (zind)*nX*nY + (yind+1)*nX + (xind+1);
                    index[4] = (zind+1)*nX*nY + (yind)*nX + (xind);
                    index[5] = (zind+1)*nX*nY + (yind)*nX + (xind+1);
                    index[6] = (zind+1)*nX*nY + (yind+1)*nX + (xind);
                    index[7] = (zind+1)*nX*nY + (yind+1)*nX + (xind+1);
followed by the call to AssignEight() using index[].The total number of points in this dataset is more than 2^31 (1536*1536*1297 = 3,060,006,912), causing negative values due to overflow in the index[] array. Which then causes out-of-bounds access in of array[] in AssignEight<T>.I'm not very familiar with the VisIt sources, so my analysis might be wrong. Because it's a bit surprising that a dataset that isn't extremely large would trigger this kind of bug. Surely, VisIt is regularly used for much larger datasets? Or is the fact that this is an unstructured grid the exception?
Paul Melis wrote:The total number of points in this dataset is more than 2^31 (1536*1536*1297 = 3,060,006,912), causing negative values due to overflow in the index[] array. Which then causes out-of-bounds access in of array[] in AssignEight<T>.I'm not very familiar with the VisIt sources, so my analysis might be wrong. Because it's a bit surprising that a dataset that isn't extremely large would trigger this kind of bug. Surely, VisIt is regularly used for much larger datasets? Or is the fact that this is an unstructured grid the exception?Yes, but ordinarily when VisIt is processing a large dataset, that dataset is broken into pieces by the data producer (e.g. simulation code) such that no one piece is O(max-int) number of elements. Different ranks operate on different pieces from bytes on disk to pixels in a framebuffer and then those framebuffers are composited together in parallel to produce the final image. Or, in the case of VR, rays are traced in parallel through those pieces.It turns out it is kinda rare to encounter single blocks of the size you are dealing with here. That said, it seems like switching to a 64 bit integer type here or even a double (which is 52 bits of integer precision) might work.
Mark Miller wrote:Paul Melis wrote:The total number of points in this dataset is more than 2^31 (1536*1536*1297 = 3,060,006,912), causing negative values due to overflow in the index[] array. Which then causes out-of-bounds access in of array[] in AssignEight<T>.I'm not very familiar with the VisIt sources, so my analysis might be wrong. Because it's a bit surprising that a dataset that isn't extremely large would trigger this kind of bug. Surely, VisIt is regularly used for much larger datasets? Or is the fact that this is an unstructured grid the exception?Yes, but ordinarily when VisIt is processing a large dataset, that dataset is broken into pieces by the data producer (e.g. simulation code) such that no one piece is O(max-int) number of elements. Different ranks operate on different pieces from bytes on disk to pixels in a framebuffer and then those framebuffers are composited together in parallel to produce the final image. Or, in the case of VR, rays are traced in parallel through those pieces.Sure, I understand that. However, I expect that producing data in multiple files is avoided as long as the post-processing tools being used provide sufficient performance on a single block (and "sufficient performance" becomes relative to the effort required to add parallel writing to a simulation code). Especially 3D volumes can be quite large when rendered on a single high-end GPU these days. You'd be surprised how many users do not use the parallel rendering features of VisIt (or ParaView for that matter) because it takes extra effort to get working.It turns out it is kinda rare to encounter single blocks of the size you are dealing with here. That said, it seems like switching to a 64 bit integer type here or even a double (which is 52 bits of integer precision) might work.One of the issues here is that there is an implicit assumption about the maximum size of the dataset, which isn't checked or enforced. Even when using 64-bit integers there will be a limit, which is fine in itself, as long as that limit is clear. So is the current limit 2^31 points per volume dataset?I will ask the user if writing multiple smaller pieces is an option.
Paul Melis wrote:I will ask the user if writing multiple smaller pieces is an option.The user's reaction:The advantage of using HDF5 is that we can output all the data from different cores to the same file automatically. Actually the simulation that carried out the dataset I gave you was running on 800 cores.This actually seems quite reasonable. Having a single result file is a lot better than multiple smaller ones, in terms of pre-processing with a diverse set of tools. Are there any tricks in Visit to be able to pick a spatial subset of the dataset, to overcome the overflow bug here? I know Xdmf allows picking complete HDF5 datasets, but I haven't looked into picking a subset of a dataset.
The gen_dummy.py program is quite helpful! Unfortunately, for me it generates non-zero data only for a tiny portion of the whole dataset. It is zero almost everywhere except in a thin region near z==0. But, that is fine too.Although in principle I agree VisIt should be able to digest a single block this size, there are unfortunately a slew of places in VisIt that won't handle this data for probably a similar reason. I am loathe to fix this one stack trace only to find out something downstream of it also won't work for similar reason. Its a whole set of dominoes that has to be touched to ensure that this one pipline will run. And, another set for a different pipeline. BTW, when I plot the mesh plot on this data set, it is so dense in mesh lines, nothing else can be seen.I think the best we can do for now is to issue an error message regarding size and then avert any attempts to plot objects of this size as Paul suggested in an earlier post. We'll have to file a different ticket to fix this issue throughout VisIt.
I have resolved this issue in a very unsatisfying way. I've added logic to TransformManager to detect blocks too big and likely to cause integer indexing arithmetic overflow and then basically set these blocks to empty. The only thing attractive about this solution is that it will avert total crashes.Another ticket has been entered to solve the problem correctly..audit use of int everywhere a size_t should be used.
